loglist2/logfilter.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loglist2
+
+import (
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/x509"
+)
+
+// LogRoots maps Log-URLs (stated at LogList) to the list of their accepted
+// root-certificates.
+type LogRoots map[string][]*x509.Certificate
+
+// SelectQualified creates a new LogList containing only qualified logs from
+// the original.
+func (ll *LogList) SelectQualified() LogList {
+	var active LogList
+	active.Operators = make(map[string]*Operator)
+	// Keep all the operators but filter Logs.
+	for opName, op := range ll.Operators {
+		active.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
+		copy(active.Operators[opName].Email, op.Email)
+		for logName, l := range op.Logs {
+			if l.State.String() != "Qualified" {
+				continue
+			}
+			active.Operators[opName].Logs[logName] = l
+		}
+	}
+	return active
+}
+
+// Compatible creates a new LogList containing only the logs of original
+// LogList that are compatible with the provided cert-chain, according to
+// the passed in collection of per-log roots. Logs that are missing from
+// the collection are treated as always compatible and included, even if
+// an empty cert chain is passed in.
+// Cert-chain when provided is expected to be full: ending with CA-cert.
+func (ll *LogList) RootCompatible(rootedChain []*x509.Certificate, roots LogRoots) LogList {
+	var compatible LogList
+	// Keep all the operators.
+	compatible.Operators = make(map[string]*Operator)
+	for opName, op := range ll.Operators {
+		compatible.Operators[opName] = &Operator{Email: make([]string, len(op.Email)), Logs: make(map[string]*Log)}
+		copy(compatible.Operators[opName].Email, op.Email)
```

Comment:
As before, you're copying some things but sharing others.

Reply:
Done.